### PR TITLE
Init: add backend frontend verifier component

### PR DESCRIPTION
**Pull request by yishangzhang**

### CODE REVIEW

File: verifier/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
+FROM ubuntu:20.04
+
+LABEL maintainer="yishangzhang <Yishang.Zhang@linux.alibaba.com>"
+
+SHELL ["/bin/bash", "-c"]
+RUN apt-get update && \
```
Comment (hustliyilin) [inline verifier/Dockerfile]: 1. change `apt-get update` as `apt-get update -y`
2. change `rm -rf /var/lib/apt/lists/*` as `rm -rf /var/lib/apt/lists`

File: verifier/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    tar zxvf ibmtpm1682.tar.gz && \
+    cd src && make install  
+
+RUN mkdir /root/tpm2 && \
+    cd /root/tpm2 && \
+    wget https://versaweb.dl.sourceforge.net/project/ibmtpm20tss/ibmtss1.6.0.tar.gz && \
+    tar zxvf ibmtss1.6.0.tar.gz
+
+RUN for i in {1..5}; do \
+        git clone https://github.com/kgoldman/acs.git /root/acs && \
```
Comment (hustliyilin) [inline verifier/Dockerfile]: add a judgement about whether the acs is ready before executing `git clone`

File: verifier/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        break; \
+        sleep 5; \
+    done
+
+WORKDIR /usr/include 
+RUN ln -s json-c json
+
+
```
Comment (hustliyilin) [inline verifier/Dockerfile]: delete this empty line

File: frontend/frontend.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ./configure --prefix=/usr --enable-debug --target-list=x86_64-softmmu --enable-kvm --enable-virtfs --enable-slirp
+    make -j$(nproc)
+    make install
+    cd ..
+else
+    echo "qemu is installed"
+fi
+
+
+
```
Comment (hustliyilin) [inline frontend/frontend.sh]: delete the empty lines(just keep onw empty line is ok).

File: frontend/frontend.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+password=$(openssl passwd -1 123456)
+
+# Mount VM image
+modprobe nbd max_part=63
+qemu-nbd -c /dev/nbd0 focal-server-cloudimg-amd64.img
+sleep 5
+mount /dev/nbd0p1 /mnt
+
+# Update shadow file with password hash
+sed -i "s#^root:.*#root:$password:18472:0:99999:7:::#" /mnt/etc/shadow
```
Comment (hustliyilin) [inline frontend/frontend.sh]: leaking passwords is risky.

File: frontend/frontend.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+if [ -f "./focal-server-cloudimg-amd64.img" ]
+then
+    echo "File exists"
+else
+    wget https://cloud-images.ubuntu.com/focal/20230420/focal-server-cloudimg-amd64.img
+fi
+
+
+# Generate password hash
+password=$(openssl passwd -1 123456)
```
Comment (hustliyilin) [inline frontend/frontend.sh]: leaking passwords is risky.

File: frontend/frontend.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+mount /dev/nbd0p1 /mnt
+
+# Update shadow file with password hash
+sed -i "s#^root:.*#root:$password:18472:0:99999:7:::#" /mnt/etc/shadow
+
+# Unmount VM image
+umount /mnt
+qemu-nbd -d /dev/nbd0
+
+
```
Comment (hustliyilin) [inline frontend/frontend.sh]: ditto

File: backend/backend.sh [hunk 8]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+
+apt-get update
```
Comment (hustliyilin) [inline backend/backend.sh]: change to `apt-get update -y`

File: backend/backend.sh [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+rm -rf /var/lib/apt/lists/*
+
+rm -rf /root/tpm2  /root/acs  /root/startup.sh /root/swtpm
+
+mkdir /root/tpm2 && \
+    cd /root/tpm2 && \
+    wget https://versaweb.dl.sourceforge.net/project/ibmtpm20tss/ibmtss1.6.0.tar.gz && \
+    tar zxvf ibmtss1.6.0.tar.gz
+
+# 多次尝试克隆 acs 代码库
```
Comment (hustliyilin) [inline backend/backend.sh]: Please notes in English

File: backend/backend.sh [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+cd /root/swtpm/libtpms && ./autogen.sh --with-tpm2 --with-openssl --prefix=/usr && \
+make -j$(nproc) && make install
+
+
+#--enable-debug
+cd /root/swtpm/swtpm && ./autogen.sh --prefix=/usr   --with-gnutls  --without-seccomp --with-tpm2  --with-openssl && \ 
+make -j$(nproc) && make install
+
+
+# 创建符号链接
```
Comment (hustliyilin) [inline backend/backend.sh]: ditto

File: backend/backend.sh [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ln -s /usr/include/json-c /usr/include/json
+
+# 编译 acs
```
Comment (hustliyilin) [inline backend/backend.sh]: ditto

File: backend/backend.sh [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    echo 'swtpm_setup --tpmstate /tmp/mytpm1 --create-ek-cert --create-platform-cert  --tpm2 --display' >> /root/startup.sh && \
+    echo 'swtpm socket --tpmstate dir=/tmp/mytpm1 --ctrl type=unixio,path=/tmp/mytpm1/swtpm-sock --log level=20 --tpm2 --server type=tcp,port=2321 &' >> /root/startup.sh && \
+    echo 'swtpm_ioctl -i --unix /tmp/mytpm1/swtpm-sock' >> /root/startup.sh && \
+    echo 'swtpm_bios --tpm2 --tcp :2321' >> /root/startup.sh && \
+    echo 'cd /root/tpm2/acs && ./clientenroll -alg rsa -vv -ho localhost  -co /root/tpm2/acs/akcert.pem' >> /root/startup.sh 
+
+
+
+
+# 设置环境变量
```
Comment (hustliyilin) [inline backend/backend.sh]: ditto

File: backend/backend.sh [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+source ~/.bashrc && chmod +x /root/startup.sh
+
+# 切换到 acs 代码库并编译 clientenroll
```
Comment (hustliyilin) [inline backend/backend.sh]: ditto

File: backend/backend.sh [hunk 14]
```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+apt-get update -y
+#need in swtpm 
+apt  install -y autoconf-archive libcmocka0 libcmocka-dev procps iproute2 build-essential git pkg-config gcc libtool automake libssl-dev uthash-dev autoconf doxygen libjson-c-dev libini-config-dev libcurl4-openssl-dev libltdl-dev expect python3-yaml libglib2.0-dev dbus  libtasn1-6-dev libjson-glib-dev gawk socat libseccomp-dev uuid-dev udev libusb-1.0-0-dev libgnutls28-dev gnutls-bin
+#need in acs
+apt  install -y libjson-c4  libjson-c-dev mariadb-server libmysqlclient-dev git make gcc wget && rm -rf /var/lib/apt/lists
+# Remove existing directories and files
+rm -rf /root/tpm2  /root/acs  /root/startup.sh /root/swtpm
```
Comment (fengyuanyu1) [inline backend/backend.sh]: I think this removing is not necessary.

File: frontend/frontend.sh (hunk 6 above)
Comment (fengyuanyu1) [reply frontend/frontend.sh]: @yishangzhang It is reasonable to delegate the action of setting password to the user.

File: verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b
Comment (fengyuanyu1) [inline verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b:1]: It is better to put the patch in the `hacks` folder. You need to sign this patch with the developer information and delete the useless comments.

**Review by fengyuanyu1 — CHANGES_REQUESTED** [review]
The directory of this repository should be as follows:
```
├─acs
├─libtpms
├─swtpm
├─qemu
└─hacks
      └─acs
           └─xxx.patch
```
You can use the `git submodule` to achieve this target, then you can replace the operation of download or clone to the `git submodule init`

File: verifier/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,58 @@
+FROM ubuntu:20.04
+
+LABEL maintainer="yishangzhang <Yishang.Zhang@linux.alibaba.com>"
+
+SHELL ["/bin/bash", "-c"]
```
Comment (haosanzi) [inline verifier/Dockerfile]: No need to add line 5

File: verifier/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    tar zxvf ibmtss1.6.0.tar.gz
+
+RUN for i in {1..5}; do \
+        git clone https://github.com/kgoldman/acs.git /root/acs && \
+        break; \
+        sleep 5; \
+    done
+
+WORKDIR /usr/include 
+RUN ln -s json-c json
```
Comment (haosanzi) [inline verifier/Dockerfile]: please remove line 27-28 to lin 9 and use `ln -s /usr/include/json-c /usr/include/json `

File: verifier/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    echo 'export LD_LIBRARY_PATH=/root/tpm2/utils:/root/tpm2/utils12' >> /root/startup.sh && \
+    echo 'export TPM_DATA_DIR=/root/tpm2' >> /root/startup.sh && \
+    echo 'export TPM_SERVER_TYPE=mssim' >> /root/startup.sh&& \
+    echo 'export ACS_PORT=2323' >> /root/startup.sh && \   
+    cat /root/startup.sh >> ~/.bashrc && \
+    echo "tpm_server &" >>  /root/startup.sh && \
+    echo "/root/tpm2/utils/powerup  &" >>  /root/startup.sh && \
+    echo "/root/tpm2/utils/startup  &" >>  /root/startup.sh && \
+    echo "service mysql start" >>  /root/startup.sh && \
+    echo "/root/tpm2/acs/server -v -root ../utils/certificates/rootcerts.txt -imacert /root/tpm2/acs/imakey.der" >> /root/startup.sh
```
Comment (haosanzi) [inline verifier/Dockerfile]: Can you use a startup.sh file and just `COPY startup.sh /root`

File: verifier/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    
+RUN source ~/.bashrc && chmod +x /root/startup.sh
+COPY *patch* /root/acs/  
+
+WORKDIR /root/acs
+RUN git checkout abdcc9592ca5d12ae7d2146eed9f92c48e5e592b  && \
+    git apply acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b && \
+    mv /root/acs/acs /root/tpm2/acs && \
```
Comment (haosanzi) [inline verifier/Dockerfile]: Why not just place `/root/acs/acs` to `/root/tpm2/acs`

File: verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+-					 &rootFileCount,
+-					 listFilename,
+-					 vverbose);
+-    }
+-    if (rc == 0) {
+-	if (vverbose)
+-	    printf("validateEkCertificate: Validate the client EK certificate against the root\n");
+-    }
++
++	//TODO : add  ca pubkey in listFilename ,which enable	verifyCertificate
```
Comment (haosanzi) [inline verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b]: please change to `// TODO : add ca pubkey in listFilename, which enable verifyCertificate`

File: verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b [hunk 20]
```diff
@@ -0,0 +1,56 @@
+diff --git a/acs/server.c b/acs/server.c
+index 25f54f4..0daba45 100644
+--- a/acs/server.c
++++ b/acs/server.c
+@@ -5325,24 +5325,29 @@ static uint32_t validateEkCertificate(TPMT_PUBLIC *ekPub,	/* output */
+     for (i = 0 ; i < MAX_ROOTS ; i++) {
+ 	rootFilename[i] = NULL;    				/* for free @1 */
```
Comment (haosanzi) [inline verifier/acs.patch.abdcc9592ca5d12ae7d2146eed9f92c48e5e592b]: Is this an unfinished patch? Just delete some code?

File: frontend/frontend.sh [hunk 21]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+
+# Install dependencies
+apt-get update && apt-get install -y libglib2.0-dev libfdt-dev libpixman-1-dev zlib1g-dev ninja-build libcap-ng-dev libslirp-dev
+
+
+# Install qemu
```
Comment (haosanzi) [inline frontend/frontend.sh]: please use a variable to represent the version number in order to maintain when we use to change the version, we just to change the variable.

File: frontend/frontend.sh [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+-device tpm-tis,tpmdev=tpm0
+
+qemu-system-x86_64 -enable-kvm  -smp 8 -m 10G  -nographic \
+-netdev user,id=vmnic -net user,hostfwd=tcp::10021-:22 \
+-net nic,model=e1000 \
+-device e1000,netdev=vmnic,romfile= \
+-hda focal-server-cloudimg-amd64.img \
+-chardev socket,id=chrtpm,path=/tmp/mytpm1/swtpm-sock \
+-tpmdev emulator,id=tpm0,chardev=chrtpm \
+-device tpm-tis,tpmdev=tpm0
```
Comment (haosanzi) [inline frontend/frontend.sh]: line 54 is same as line63?

File: backend/backend.sh [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+apt-get update
+
+
+#need in swtpm 
+apt -y install autoconf-archive libcmocka0 libcmocka-dev procps iproute2 build-essential git pkg-config gcc libtool automake libssl-dev uthash-dev autoconf doxygen libjson-c-dev libini-config-dev libcurl4-openssl-dev libltdl-dev expect python3-yaml libglib2.0-dev dbus  libtasn1-6-dev libjson-glib-dev gawk socat libseccomp-dev uuid-dev udev libusb-1.0-0-dev libgnutls28-dev gnutls-bin
+
+
+#need in acs
+apt-get install -y libjson-c4  libjson-c-dev mariadb-server libmysqlclient-dev git make gcc wget 
+rm -rf /var/lib/apt/lists/*
```
Comment (haosanzi) [inline backend/backend.sh]: `gcc ` is installed in line 8 and line 12. please check the same problem

File: frontend/frontend.sh (hunk 22 above)
Comment (yishangzhang) [reply frontend/frontend.sh]: the file was outdated

File: backend/backend.sh [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+rm -rf /root/tpm2  /root/acs  /root/startup.sh /root/swtpm
+
+mkdir /root/tpm2 && \
+    cd /root/tpm2 && \
+    wget https://versaweb.dl.sourceforge.net/project/ibmtpm20tss/ibmtss1.6.0.tar.gz && \
```
Comment (haosanzi) [inline backend/backend.sh]: please use `wget -P /toot/tpm2 https://versaweb.dl.sourceforge.net/project/ibmtpm20tss/ibmtss1.6.0.tar.gz` to replace line 18 and 19

File: backend/backend.sh [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    tar zxvf ibmtss1.6.0.tar.gz
+
```
Comment (haosanzi) [inline backend/backend.sh]: please add remove ibmtss1.6.0.tar.gz command to save disk space.

File: backend/backend.sh (hunk 23 above)
Comment (yishangzhang) [reply backend/backend.sh]: Appearing in two places is to quickly find all the dependencies for the subsequent replacement of the component location

File: backend/backend.sh [hunk 26]
```diff
@@ -1,55 +1,47 @@
...
 mkdir /root/tpm2 && \
     cd /root/tpm2 && \
     wget https://versaweb.dl.sourceforge.net/project/ibmtpm20tss/ibmtss1.6.0.tar.gz && \
     tar zxvf ibmtss1.6.0.tar.gz
-
-# 多次尝试克隆 acs 代码库
+# Attempt to clone the acs repository up to 5 times with a 5 second sleep in between each attempt
 for i in {1..5}; do \
-    git clone https://github.com/kgoldman/acs.git /root/acs && \
+    [ ! -d "/root/acs" ] git clone https://github.com/kgoldman/acs.git /root/acs && \
```
Comment (haosanzi) [inline backend/backend.sh]: When multi commit in one pr, there is no need to submit a new commit to fix the problem commit. just use `git rebase ` to fix the problem commit.

File: frontend/frontend.sh [hunk 27]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+export DEBIAN_FRONTEND=noninteractive
+#Step 1. Install dependencies
```
Comment (fengyuanyu1) [inline frontend/frontend.sh]: Better to put the space between `#` and following contents.

File: frontend/frontend.sh [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+apt-get update -y && apt install -y libglib2.0-dev libfdt-dev libpixman-1-dev zlib1g-dev ninja-build libcap-ng-dev libslirp-dev
+
+#Step 2. Install qemu
```
Comment (fengyuanyu1) [inline frontend/frontend.sh]: ditto

**Comment by fengyuanyu1** [comment]
Thx for your work @yishangzhang , there are only several problems.

1. The directory can be better organized as follows:

```
├─acs
├─libtpms
├─swtpm
├─qemu
├─scripts
│    ├─backend
│    │      └─backend.sh
│    ├─frontend
│    │      └─frontend.sh
│    └─verifier
│            └─Dockerfile
└─patches or hacks
      ├─acs
      │     └─xxx.patch
      ├─qemu
      │     └─xxx.patch
      └─ ....
```
This change may influence the scripts, you need to check them.

2. This patch [patchs/acs/0001-del-ekcert-verify.patch](https://github.com/inclavare-containers/confidential-vtpm/pull/4/files#diff-c27ad1425b73449ebcb2daeecc30e93af554e3e017786d17812215c8c3e8c0bd) has many comment lines, you can regenerate one without them.

File: scripts/backend/backend.sh [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+
+# Step 8. Compile tpm2 tss
+cd $CURRENT_DIR/tpm2/utils
+make -f makefiletpm20 && \
+    echo export CPATH=$CURRENT_DIR/tpm2/utils >> $CURRENT_DIR/startup.sh && \
+    echo export LD_LIBRARY_PATH=$CURRENT_DIR/tpm2/utils:$CURRENT_DIR/tpm2/utils12“ >> $CURRENT_DIR/startup.sh && \
+    echo export TPM_DATA_DIR=$CURRENT_DIR/tpm2 >> $CURRENT_DIR/startup.sh && \
+    echo export TPM_SERVER_TYPE=mssim >> $CURRENT_DIR/startup.sh && \
+    echo export ACS_PORT=2323 >> $CURRENT_DIR/startup.sh && \
+    cat $CURRENT_DIR/startup.sh >> ~/.bashrc && \
```
Comment (fengyuanyu1) [inline scripts/backend/backend.sh]: Maybe we can replace the compilation using startup.sh and .bashrc by using the .bashrc only.

**Review by fengyuanyu1 — APPROVED** [review]
LGTM!

File: scripts/backend/backend.sh [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+apt-get update -y
+
+# Step 1. Install swtpm requirements
+apt install -y autoconf-archive libcmocka0 libcmocka-dev procps iproute2 build-essential git pkg-config gcc libtool automake libssl-dev uthash-dev autoconf doxygen libjson-c-dev libini-config-dev libcurl4-openssl-dev libltdl-dev expect python3-yaml libglib2.0-dev dbus  libtasn1-6-dev libjson-glib-dev gawk socat libseccomp-dev uuid-dev udev libusb-1.0-0-dev libgnutls28-dev gnutls-bin
+
+# Step 2. Install acs requirements
+apt install -y libjson-c4 libjson-c-dev mariadb-server libmysqlclient-dev git make gcc wget && rm -rf /var/lib/apt/lists
+
+# Step 3. download submoudle
+CURRENT_DIR=$(pwd)
```
Comment (hustliyilin) [inline scripts/backend/backend.sh:12]: Please use `WORKDIR` to replace the `CURRENT_DIR`. And you can remove all `CURRENT_DIR` in this Dockerfile

File: scripts/frontend/frontend.sh [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+# Step 6.Launch VM
+qemu-system-x86_64 -enable-kvm -m 2048  -nographic \
+-netdev user,id=vmnic -net user,hostfwd=tcp::10021-:22 \
+-net nic,model=e1000 \
+-device e1000,netdev=vmnic,romfile= \
+-hda  focal-server-cloudimg-amd64.img \
+-chardev socket,id=chrtpm,path=/tmp/mytpm2/swtpm-sock \
+-tpmdev emulator,id=tpm0,chardev=chrtpm \
+-device tpm-tis,tpmdev=tpm0
+
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh:53]: delete the extra empty line

File: scripts/frontend/frontend.sh [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+sed -i "s#^$username:.*#$username:$password:18472:0:99999:7:::#" /mnt/etc/shadow
+    # Unmount VM image
+umount /mnt
+qemu-nbd -d /dev/nbd0
+
+# Step 5. Launch swtpm
+mkdir /tmp/mytpm2
+swtpm socket --tpmstate dir=/tmp/mytpm2 \
+--ctrl type=unixio,path=/tmp/mytpm2/swtpm-sock \
+--log level=20 --tpm2 &  >/dev/null 2>&1
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: add an empty line

File: scripts/frontend/frontend.sh [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+then
+    echo "File exists"
+else
+    wget https://cloud-images.ubuntu.com/focal/20230420/focal-server-cloudimg-amd64.img
+fi
+
+# Step 4. config VM name and password
+read -p "Enter username: " username
+read -p "Enter password: " password
+    # Generate password hash
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: Please fix the indentation error

File: scripts/frontend/frontend.sh [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+password=$(openssl passwd -1 $password)
+    # Mount VM image
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: ditto

File: scripts/frontend/frontend.sh [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+modprobe nbd max_part=63
+qemu-nbd -c /dev/nbd0 focal-server-cloudimg-amd64.img
+sleep 5
+mount /dev/nbd0p1 /mnt
+    # Update shadow file with password hash
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: ditto

File: scripts/frontend/frontend.sh [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+sed -i "s#^$username:.*#$username:$password:18472:0:99999:7:::#" /mnt/etc/shadow
+    # Unmount VM image
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: ditto

File: scripts/frontend/frontend.sh [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+umount /mnt
+qemu-nbd -d /dev/nbd0
+
+# Step 5. Launch swtpm
+mkdir /tmp/mytpm2
+swtpm socket --tpmstate dir=/tmp/mytpm2 \
+--ctrl type=unixio,path=/tmp/mytpm2/swtpm-sock \
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: Please fix the indentation error. It should be

```shell
swtpm socket --tpmstate dir=/tmp/mytpm2 \
             --ctrl type=unixio,path=/tmp/mytpm2/swtpm-sock \
             --log level=20 --tpm2 &  >/dev/null 2>&1 
```

File: scripts/frontend/frontend.sh [hunk 38]
```diff
@@ -0,0 +1,70 @@
...
+--log level=20 --tpm2 &  >/dev/null 2>&1 
+# Step 6.Launch VM
+qemu-system-x86_64 -enable-kvm -m 2048  -nographic \
+-netdev user,id=vmnic -net user,hostfwd=tcp::10021-:22 \
```
Comment (hustliyilin) [inline scripts/frontend/frontend.sh]: ditto

**Review by hustliyilin — APPROVED** [review]
LGTM

File: startup.sh [hunk 39]
```diff
@@ -1,14 +0,0 @@
...
-export LD_LIBRARY_PATH=/root/tpm2/utils:/root/tpm2/utils12“
-export TPM_DATA_DIR=/root/tpm2
-export TPM_SERVER_TYPE=mssim
-export ACS_PORT=2323
-
-mkdir /tmp/mytpm1
-swtpm_setup --tpmstate /tmp/mytpm1 --create-ek-cert --create-platform-cert  --tpm2 --display
-swtpm socket --tpmstate dir=/tmp/mytpm1 --ctrl type=unixio,path=/tmp/mytpm1/swtpm-sock --log level=20 --tpm2 --server type=tcp,port=2321 &
-sleep 3 
-swtpm_ioctl -i --unix /tmp/mytpm1/swtpm-sock
```
Comment (haosanzi) [inline startup.sh]: Why delete this file? If you do Not need this file, just do not add in previous commit https://github.com/inclavare-containers/confidential-vtpm/pull/4/commits/49566b33d6c59cc6a6c71baca0991472a6320da7

Comment (yishangzhang) [reply startup.sh]: I have already deleted this file in the previous commit

**Review by haosanzi — APPROVED** [review]

**Review by hustliyilin — APPROVED** [review]

**Pull request merged**